### PR TITLE
Addition of matrix product

### PR DESCRIPTION
**Pull request by YvanGuifo**
It contains :
- A method allowing to check that the matrices which one wishes to multiply have the same shape.
- A method to calculate the Hadamard product
- And test methods for this Hadamard product

### CODE REVIEW

File: src/Math-Matrix/PMNDArray.class.st [hunk 1]
```diff
@@ -41,7 +41,7 @@ ifTrue:[
 { #category : #comparing }
 PMNDArray >> = anArray [
```
Comment (SergeStinckwich) [inline src/Math-Matrix/PMNDArray.class.st:43]: Maybe use ``aPMNDArray`` instead of ``anArray``, because this is not an instance of ``Array`` class.

**Review by SergeStinckwich — CHANGES_REQUESTED** [review]

File: src/Math-Matrix/PMNDArray.class.st (hunk 1 above)
Comment (YvanGuifo) [reply src/Math-Matrix/PMNDArray.class.st:43]: Ok. Thanks

**Review by khinsen — COMMENTED** [review]
This looks OK as an implementation of the Hadamard product.But is that what we want? I'd expect to have element-wise application for any binary method (or two-argument code block), not just multiplication. So instead of
```
   array1 hadamardProduct: array2
```
I'd expect to write
```
   array1 with: array2 collect: #'*'
```

**Comment by YvanGuifo** [comment]
Thank you @khinsen.  Need to clarify. When you talk about binary operation. Why look for other operators when it is a product that we perform.

**Comment by khinsen** [comment]
It's a matter of interface design. Years of practice with APL, NumPy, and derived languages have shown that elementwise application of binary operations is very frequent. Products are nothing special, they are not used more often than addition and substraction. So, rather than introducing a specific operation for the product, I'd introduce a general one for any two-argument function.

APL and NumPy go even further and use the plain binary operators for elementwise combination. The Hadamard product of `a` and `b` is simply `a * b` in NumPy (or `a × b` in APL).

File: src/Math-Matrix/PMNDArray.class.st [hunk 2]
```diff
@@ -97,6 +97,21 @@ PMNDArray >> fromNestedArray: aFlatArray withShape: aShape [
 	shape ifNotEmpty: [ self updateStrides]
 ]
 
+{ #category : #operation }
+PMNDArray >> hadamardProduct: aPMNDArray [
+	(self hasSameShapeAs: aPMNDArray)
+	ifTrue:[
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMNDArray.class.st:103]: This is the normal flow, is that correct? If so, consider moving this out of the block and have run after the check. `ifFalse` becomes a guard clause. The code, in my opinion, then becomes a little more maintainable.

File: src/Math-Matrix/PMNDArray.class.st [hunk 3]
```diff
@@ -97,6 +97,21 @@ PMNDArray >> fromNestedArray: aFlatArray withShape: aShape [
...
+		^ self class new array: ((self asArray) with: (aPMNDArray asArray) collect:[:a :b| a*b])
+		withShape: self shape.
+		]
+	ifFalse:[ShapeMismatch signal ]
+]
+
+{ #category : #testing }
+PMNDArray >> hasSameShapeAs: aPMNDArray [
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMNDArray.class.st:111]: Really nice name!

File: src/Math-Matrix/PMNDArrayTest.class.st [hunk 4]
```diff
@@ -100,6 +100,53 @@ PMNDArrayTest >> testFromNestedArray [
...
+					).
+					
+				
+			
+		
+	
+]
+
+{ #category : #tests }
+PMNDArrayTest >> testHadamardProductWithDifferentShapes [
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMNDArrayTest.class.st:143]: In the name, my suggestion is to emphasise in the test method name, the fact that it fails.

**Comment by SergeStinckwich** [comment]
> It's a matter of interface design. Years of practice with APL, NumPy, and derived languages have shown that elementwise application of binary operations is very frequent. Products are nothing special, they are not used more often than addition and substraction. So, rather than introducing a specific operation for the product, I'd introduce a general one for any two-argument function.
> 
> APL and NumPy go even further and use the plain binary operators for elementwise combination. The Hadamard product of `a` and `b` is simply `a * b` in NumPy (or `a × b` in APL).

I think the suggestion of Konrad, is to introduce the elementwise method: ``with:collect:`` and then refactor ``hadamardProduct`` accordingly.

File: src/Math-Matrix/PMNDArray.class.st (hunk 2 above)
Comment (SergeStinckwich) [reply src/Math-Matrix/PMNDArray.class.st:103]: Agree, something like:
```
(self hasSameShapeAs:aPMNDArray)
ifFalse:[ShapeMismatch signal].
^ self class new array: ((self asArray) with: (aPMNDArray asArray) collect:[:a :b| a*b])
		withShape: self shape.
```

```

File: src/Math-Matrix/PMNDArrayTest.class.st (hunk 4 above)
Comment (YvanGuifo) [reply src/Math-Matrix/PMNDArrayTest.class.st:143]: Ok. Thanks

**Review by hemalvarambhia — APPROVED** [review]
All good.

File: src/Math-Matrix/PMNDArray.class.st [hunk 5]
```diff
@@ -97,6 +97,18 @@ PMNDArray >> fromNestedArray: aFlatArray withShape: aShape [
 	shape ifNotEmpty: [ self updateStrides]
 ]
 
+{ #category : #operation }
+PMNDArray >> hadamardProduct: aPMNDArray [
+	^ self with: aPMNDArray collect: [:a :b| a*b]
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMNDArray.class.st:102]: Spacing would help me in `a*b`

File: src/Math-Matrix/PMNDArrayTest.class.st [hunk 6]
```diff
@@ -100,6 +100,53 @@ PMNDArrayTest >> testFromNestedArray [
 	
 ]
 
+{ #category : #tests }
+PMNDArrayTest >> testHadamardProduct [
+|t1 t2|
+ t1 := PMNDArray fromNestedArray:  #(
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMNDArrayTest.class.st:106]: Suggestion: In the tests, see if you can give these better names

File: src/Math-Matrix/PMNDArrayTest.class.st [hunk 7]
```diff
@@ -100,6 +100,53 @@ PMNDArrayTest >> testFromNestedArray [
...
+					#( #( 9 10) #( 11 12))
+					).
+					
+ t2 := PMNDArray fromNestedArray:
+				 #( 
+					#( #( 1 3 ) #( 2 1)) 
+					#( #( 0 6 ) #( 3 8))
+					#( #( 3 5) #( 1 10))
+					).
+self assert: (t1 hadamardProduct: t2) equals: (PMNDArray fromNestedArray:
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMNDArrayTest.class.st:129]: Suggestion: extract the expected outcome to an intention revealing variable e.g. `expectedHadamardProduct`

**Comment by khinsen** [comment]
> I think the suggestion of Konrad, is to introduce the elementwise method: `with:collect:` and then refactor `hadamardProduct` accordingly.

Yes, exactly. Although I also wonder if we need the method `hadamardProduct` at all.

File: src/Math-Matrix/PMNDArrayTest.class.st (hunk 7 above)
Comment (YvanGuifo) [reply src/Math-Matrix/PMNDArrayTest.class.st:129]: Hi @hemalvarambhia  you suggest I change `testHadamardProduct` to `expectedHadamardProduct`

Comment (hemalvarambhia) [reply src/Math-Matrix/PMNDArrayTest.class.st:129]: No, and I apologise for lack of clarity on my part. The `PMNDArray fromNestedArray...` part of the assertion should be extracted to a local variable called `expectedHadamardProduct`

Comment (YvanGuifo) [reply src/Math-Matrix/PMNDArrayTest.class.st:129]: Thank you @hemalvarambhia but I do not really understand the interest of defining a local variable in this test

Comment (YvanGuifo) [reply src/Math-Matrix/PMNDArrayTest.class.st]: OK. No, I understand the concerns of readability and documentation mentioned

**Review by SergeStinckwich — APPROVED** [review]

**Pull request merged**